Exercicis per practicar amb regressions.

In [51]:
import pandas as pd

In [52]:
print(open('housing data description.txt').read())

1. Title: Boston Housing Data

2. Sources:
   (a) Origin:  This dataset was taken from the StatLib library which is
                maintained at Carnegie Mellon University.
   (b) Creator:  Harrison, D. and Rubinfeld, D.L. 'Hedonic prices and the 
                 demand for clean air', J. Environ. Economics & Management,
                 vol.5, 81-102, 1978.
   (c) Date: July 7, 1993

3. Past Usage:
   -   Used in Belsley, Kuh & Welsch, 'Regression diagnostics ...', Wiley, 
       1980.   N.B. Various transformations are used in the table on
       pages 244-261.
    -  Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning.
       In Proceedings on the Tenth International Conference of Machine 
       Learning, 236-243, University of Massachusetts, Amherst. Morgan
       Kaufmann.

4. Relevant Information:

   Concerns housing values in suburbs of Boston.

5. Number of Instances: 506

6. Number of Attributes: 13 continuous attributes (including "class"
                

In [53]:
columnas = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
viviendas = pd.read_csv('housing data.csv', sep=',', names=columnas)

In [54]:
viviendas.head(2)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   

   PTRATIO      B  LSTAT  MEDV  
0     15.3  396.9   4.98  24.0  
1     17.8  396.9   9.14  21.6

Lo primero que hemos hecho ha sido cargar todos los datos del dataframe `housing data.csv` en la variable viviendas y asignar los nombres de las columnas del mismo.

EX01 Crea almenys dos models de regressió diferents per intentar predir el millor possible el preu de les vivendes (MEDV) de l'arxiu adjunt.

In [55]:
import numpy as np
from sklearn.linear_model import LinearRegression

Utilizaremos dos tipos de regresión lineal múltiple, una de primer grado: modelo 1 (model) y otra polinómica de segundo grado: modelo 2 (model2).

# Modelo 1

En el primer modelo todas las variables del dataframe actúan como variables independientes (x) excepto MEDV, que actúa como variable dependiente (y).

In [56]:
x = np.array(viviendas[columnas].drop('MEDV', axis = 1))
y = np.array(viviendas.MEDV)

print(x.shape)
y.shape

(506, 13)


(506,)

Cargamos los datos para x e y, convirtiéndolos desde un dataframe (formato original) a dos arrays numpy, tal y como requiere el modelo. X es multidimensional (con 506 filas y 13 columnas), mientras y es unidimensional (con 506 filas).

In [57]:
model = LinearRegression().fit(x, y)

Ajustamos el modelo a los datos de x e y.

In [58]:
print(f"intercept: {model.intercept_}")
print(f"coefficients: {model.coef_}")

intercept: 36.459488385089955
coefficients: [-1.08011358e-01  4.64204584e-02  2.05586264e-02  2.68673382e+00
 -1.77666112e+01  3.80986521e+00  6.92224640e-04 -1.47556685e+00
  3.06049479e-01 -1.23345939e-02 -9.52747232e-01  9.31168327e-03
 -5.24758378e-01]


Obtenemos el punto de intersección y los coeficientes, que correponden a la estructura: y = b0 + b1·x1 + b2·x2 + ...+ b13·x13. En este caso: y = 36.459 - 0.108·CRIM + 0.046·ZN + ... - 0.52·LSTAT

# Modelo 2

El segundo modelo es una regresión múltiple polinomial de grado 2.

In [59]:
from sklearn.preprocessing import PolynomialFeatures

In [60]:
x_ = PolynomialFeatures(degree=2, include_bias=False).fit_transform(x)
x_

array([[6.32000000e-03, 1.80000000e+01, 2.31000000e+00, ...,
        1.57529610e+05, 1.97656200e+03, 2.48004000e+01],
       [2.73100000e-02, 0.00000000e+00, 7.07000000e+00, ...,
        1.57529610e+05, 3.62766600e+03, 8.35396000e+01],
       [2.72900000e-02, 0.00000000e+00, 7.07000000e+00, ...,
        1.54315409e+05, 1.58310490e+03, 1.62409000e+01],
       ...,
       [6.07600000e-02, 0.00000000e+00, 1.19300000e+01, ...,
        1.57529610e+05, 2.23851600e+03, 3.18096000e+01],
       [1.09590000e-01, 0.00000000e+00, 1.19300000e+01, ...,
        1.54802902e+05, 2.54955600e+03, 4.19904000e+01],
       [4.74100000e-02, 0.00000000e+00, 1.19300000e+01, ...,
        1.57529610e+05, 3.12757200e+03, 6.20944000e+01]])

Transformamos la variable x para incluir los valores de x al cuadrado, y sin incluir el término de sesgo (include_bias=False), ya que por defecto ya lo aplica a regresión lineal que ajustamos a continuación.

In [61]:
model2 = LinearRegression().fit(x_, y)

Ajustamos el modelo ...

In [62]:
print(f"intercept: {model.intercept_}")
print(f"coefficients: {model.coef_}")

intercept: 36.459488385089955
coefficients: [-1.08011358e-01  4.64204584e-02  2.05586264e-02  2.68673382e+00
 -1.77666112e+01  3.80986521e+00  6.92224640e-04 -1.47556685e+00
  3.06049479e-01 -1.23345939e-02 -9.52747232e-01  9.31168327e-03
 -5.24758378e-01]


... y de manera similar al modelo de grado 1, obtenemos los valores de intersección y coeficientes.

EX02 Compara’ls en base al MSE i al R2.

R2 en el primer modelo y segundo modelo, respectivamente:



In [63]:
print(model.score(x, y))
model2.score(x_, y)

0.7406426641094095


0.9289961714593018

Una R2 más próxima a 1 se interpreta en general como una mejor aproximación, lo cual haría al segundo modelo (polinomial) mejor que el primero en la previsión del precio de la vivienda. Por otro lado hay que tener en cuenta que también se corre el riesgo de sobreajuste del modelo a los datos, lo que podría llegar a suceder en el segundo caso con un valor de R2 de 0.92, pero esto lo comprobaremos en el ejercicio 4 con los datos de prueba que separamos (test).

MSE en el primer y segundo modelos respectivamente:

In [64]:
from sklearn.metrics import mean_squared_error
import math

y_ = model.predict(x)
y_2 = model2.predict(x_)

print(mean_squared_error(y, y_))
print(mean_squared_error(y, y_2))

21.894831181729202
5.994111690794227


Que el MSE sea de 22 significa que en promedio el modelo se equivoca en 4.69 miles de dólares (raíz de 22) en la previsión del precio de la vivienda en base a los valores de las variables.  
En este caso vemos como el error es significativamente mayor que el que se comete con un modelo cuadrático, de 2.44 miles de dólares (raíz de 6).
Es decir, 4.690 frente a 2.440 dólares de error en promedio. Mejor el modelo de grado 2 también según este segundo parámetro de medida del rendimiento.

EX03 Entrena’ls utilitzant els diferents paràmetres que admeten per intentar millorar-ne la predicció.

Utilizaremos el segundo modelo anterior, cuadrático, que era el que obtenía un mayor rendimiento, para entrenar y luego comprobar el sobreajuste o no a los datos mediante los datos de prueba (test).

In [65]:
from sklearn.model_selection import train_test_split

In [66]:
x_train, x_test, y_train, y_test = train_test_split(x_, y, test_size=0.33, random_state=4)

En primer lugar dividimos los datos en train y test.

In [67]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline

In [68]:
model3 = LinearRegression(fit_intercept=True).fit(x_train, y_train)

... ajustamos la regresión a x_train e y_train, utilizando el único parámetro de ajuste de los datos que pueden mejorar el rendimiento que actualmente existe en sklearn como parámetro.

EX04. Compara el seu rendiment emprant l’aproximació train/test o emprant totes les dades (validació interna).

In [69]:
print(model3.score(x_train, y_train))
print(model3.score(x_test, y_test))
print(model3.score(x_, y))

0.9214334838639593
0.8768105408310134
0.9026649916983766


Aquí vemos la comparativa bajo el modelo polinómico de grado 2 (model2) del rendimiento (R2) bajo entrenamiento: 0.92 - test: 0.87, y con todos los datos: 0.90. Esto muestra cuán aproximadas son las observaciones reales de las observaciones estimadas en base al modelo. Tiene cierto sentido que la R2 sea mayor con train, pues la regresión se ha hecho basándose en esos datos. Un poco inferior en test, pero todavía suficientemente alta, y algo mayor trabajando con todos los datos, al haber un número mayor de observaciones, pero menor que con train.

EX05 No facis servir la variable del nombre d'habitacions (RM) a l’hora de fer prediccions.

# Con el modelo lineal

In [70]:
x = np.array(viviendas[columnas].drop(['RM', 'MEDV'], axis = 1))

Lo primero que hacemos es excluir de x los valores de la columna RM.

In [71]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.33, random_state=4)

Partimos los datos con una proporción de 0.33 para test y 0.67 para train.

In [72]:
model4 = LinearRegression().fit(x_train, y_train)

Ajustamos el modelo lineal con los nuevos datos de entrenamiento de x.

In [73]:
print(model4.score(x_train, y_train))
print(model4.score(x_test, y_test))
print(model4.score(x, y))

0.7075550118615358
0.6530652633478382
0.6862260885468371


Finalmente obtenemos el rendimiento para train, test y con todos los valores respectivamente del modelo lineal con los nuevos valores de x (excluyendo RM).  
Vemos como el rendimiento es claramente inferior a la versión anterior, lo cuál nos hace pensar que RM tiene un alto valor explicativo en el precio de la vivienda de Boston.

# Con el modelo cuadrático

Repetimos el mismo procedimiento, pero ahora con el modelo cuadrático (polinomial o lineal de grado 2), que anteriormente nos había dado un mejor ajuste.

In [74]:
x_ = PolynomialFeatures(degree=2, include_bias=False).fit_transform(x)

In [75]:
x_train, x_test, y_train, y_test = train_test_split(x_, y, test_size=0.33, random_state=4)

In [76]:
model5 = LinearRegression().fit(x_, y)

In [77]:
print(model5.score(x_train, y_train))
print(model5.score(x_test, y_test))
print(model5.score(x_, y))

0.834947721069482
0.8840599614432167
0.8576390234166235


Con el modelo polinomial los resultados que obtenemos son mejores, como sucedía anteriormente, pero inferiores de nuevo a lo que obteníamos incluyendo la variable 'RM'.

Conclusiones:

1. La variable 'RM' tiene un cierto valor explicativo sobre el valor de la vivienda, pero no es la única, en tanto seguimos obteniendo rendimientos de entorno al 85% aún excluyéndola, utilizando el resto de variables.

2. El modelo cuadrático ajusta mejor los datos que el modelo lineal de grado uno.

Finalmente, es posible predecir con bastante exactitud el precio de la vivienda utilizando todas las variables disponibles, incluso excluyendo una de las que puede tener mayor correlación como es el número de habitaciones ('RM').
